### DISCUSSION TITLE
Project builds ok but doesn't run on Pico W

### DISCUSSION THREAD

**Discussion by JasonWheatley** [astlouys/Pico-Green-Clock#19] [Q&A]
Hi everyone,

Firstly, thank you André for this fabulous firmware! I've been running the non 'W' firmware for a while and thought maybe I'd use the PicoW to make a NTP version. Happily I see that it's already been done. Excellent!

The problem I have is my build will not run on the PicoW. The **Pico-Green-Clock.uf2.PicoW** file included with the checkout runs fine on my PicoW hardware. I get the flashing LED and debug messages about network attempt to connect and fail attempts via the serial port. I need to get my build to run as I don't want to manually enter my 63 digit random character WiFi key :-)

This is my process: 

1. Checkout the build (**git clone https://github.com/astlouys/Pico-Green-Clock.git**)
2. Uncomment the **#define PICO_W** in **Pico-Green-Clock.c**
3. Copy the contents of CMakeLists.txt.PicoW to CMakeLists.txt. 
4. Run **cmake** and then **make -j4**.
5. Connect PicoW while holding down boot button.
6. Copy **build/Pico-Green-Clock.uf2** to the RPI-RP2 folder when PicoW shows up as a drive.

Any ideas on what the problem could be would be most welcome.

Many thanks and best regards,

Jason

**Comment by astlouys** [comment]
Hi Jason,

The recipe you listed seems to be OK, but you didn't mention what the problem is... The Firmware doesn't start at all ? You don't see the time on the clock ? Or you're not able to access your network ? Have you wrote your credentials (SSID and password) in the place reserved for that for that in the code ?

Also, you must be aware that you may have problems with USB CDC with the Pico W. When the CYW43 is being initialized, USB serial communication often get lost. Is your problem related to that ?

André

Just for my curiosity, may I ask you in which country you are ?

**Reply by JasonWheatley** [reply]
Hello again André,

Thank you for your response. Apologies for not fully reporting the issue.

When I use the prebuilt image (from the checkout) **Pico-Green-Clock.uf2.PicoW**, the PicoW onboard LED flashes, the clock display goes through the expected display test and then report of hardware, software and config values. It doesn't connect to the network because it does not yet have the credentials. I also get messages received via the serial/USB port showing network connection attempts.

When I use the image that I have built, the PicoW onboard LED does not flash, there is no clock display and no messages received via serial port.

I have just updated the code to add my network credentials in the #defines of NETWORK_NAME & NETWORK_PASSWORD and also changed  **#define DST_COUNTRY** to **DST_EUROPE**. I then ran the build process again. 

I have attached my hex file to see if that helps with diagnosing the issue. I have redacted my WiFi credentials in the hex file at lines 11182 to 11185. [Pico-Clock-Green.txt](https://github.com/astlouys/Pico-Green-Clock/files/11030215/Pico-Clock-Green.txt)

No problem with revealing my location, I am in the UK :-)


Best regards,

Jason

**Reply by astlouys** [reply]
Hello Jason,

Thank you for sending those details. Unfortunately, it doesn't help me pinpointing the problem.

Did you make any other change to the source code ?  Are you still building with the #define RELEASE_VERSION ? (and not the #define DEVELOPER_VERSION). The build process doesn't give you any indication of what can be wrong ? (no error and / or warning message ?)

If you use "make -j4", I assume you're not using an IDE like Visual Studio Code (or another similar) ?

Have you visually scanned the CMakeCache.txt file to check that everything seems OK (for example, that the target platform is defined as Pico W) ? (or the CMakeError.log or CMakeOutput.log files ?)

If you copied CMakeList.txt.PicoW over CMakeList.txt at the beginning of the process and turned on the #define PICO_W in the source code, that's mostly it... Have you updated your SDK to get the latest libraries required for Pico W ? (I assume you did, otherwise, the build process would give you pages of error messages).

I know that this will not help you fixing your problem but I forgot to mention in my previous email: I know that it takes a few minutes to enter "your 63 random characters password" (smile) using the Green-Clock function that I implemented, but remember that this is a "one-time" process. Once it has been entered, it is saved to flash (in an area external to the Firmware) and will be retrieve if / when you restart the clock and this, even if you upgrade the Firmware version (I do not promise there will never be exception to that with future Firmware upgrades, but this is the way it has been thought and implemented).

I will wait for your answer in case it gives me some more ideas about the source of the problem. I may also try to build the code as you've done in the next days (I usually use Visual Studio Code as my IDE) and see if I discover something. There may be other users that could provide inputs...

The HEX file doesn't help much. You could remove your credentials from the source code and send it, but if you didn't change anything else that what you mentioned, it won't help much either.

I double-checked in the code and there are 39 characters reserved for SSID and 70 characters for password (plus the end-of-string), so the problem doesn't seem related to that. One thing I would suggest is to remove all the changes you have done and build the plain original version from GitHub to see if it works. If you get it to work, it will still be time to adapt it with your own changes,

André

**Reply by JasonWheatley** [reply]
Hi André,

I entered the SSID & password in manually on the pre-built PicoW firmware and this helped with my understanding of the system operation. I rebuilt my image with debug mode enabled and after waiting a while things started to happen. It appears the build does work it was just a case of me not RTFM. 

Although the SSID and password are correct the ntp_init() function is saying that cyw43_arch_wifi_connect_blocking() returns -7. I haven't quite figured out what that is yet. 

Looking at [this web page](https://www.i-programmer.info/programming/hardware/16031-the-picow-in-c-a-better-connect.html?start=1), I'm wondering if I need to use the function **cyw43_arch_init_with_country()** prior to the call to **cyw43_arch_wifi_connect_blocking()**? Also wondering if it needs to be **CYW43_AUTH_WPA2_MIXED_PSK** instead of **CYW43_AUTH_WPA2_AES_PSK**? Both of the previous may be an issue for UK Wi-Fi devices?

Anyhow I'll look into this more and report back when I have some results.

Thanks again for your continued support.

Best regards,

Jason

**Reply by astlouys** [reply]
Hi Jason,

Good thing that you've been able to get a successful build and finally read the guide (!!)

It seems that it's "normal" to get error messages on the "cyw43_arch_wifi_connect_blocking()" I read a few articles about this on the Internet, although the exact reason is not clear. That's why you can see that I included a retry count (up to 10 times) to make sure it will go through. I usually have to try 2 or 3 times at home (sometimes up to 4 times). If ever you find a more precise reason, I would be happy to know it !

I think adding the country would help to receive a NTP server URL near to your country, but it's not mandatory.

Hope you'll enjoy the new Firmware Version !

André

**Comment by nnmalex** [comment]
I have the same issue. Only things changed:
- Set SSID and Password
- uncomment #define PICO_W
- Copy the Correct CMakeLists

Compilation goes fine but the clock doesn't boot at all.

Commenting back the #define PICO_W makes things work (but without the NTP obviously).

**Reply by astlouys** [reply]
Hello Alex,

Same questions:

- Have you changed anything else in the source code ?
- Are you using the #define RELEASE_VERSION (and not the #define DEVELOPER_VERSION) ?
- No warning / no error at all from the compiler ?
- What is you development environment (O.S. / IDE / etc...)
- Any clue that you can see in the CMakeCache.txt or the CMakeError.log or CMakeOutput.log files ?
- Do you see the Pico W's LED blinking during the boot sequence  (through the Green-Clock casing) ?
- Did you update your SDK to cover the Pico W ?
- Have you tried to enter your SSID and Password on the pre-built version ?  Did NTP work this way ?

Jason, if you read this, maybe you could share your experience and tell us the details about what happened to you and what you changed to make things working and / or what parts of the manual you forgot to read before getting the clock to work fine... 

André

**Reply by nnmalex** [reply]
Have you changed anything else in the source code ?
- No
Are you using the #define RELEASE_VERSION (and not the #define DEVELOPER_VERSION) ?
- RELEASE_VERSION
No warning / no error at all from the compiler ?
- None at all
What is you development environment (O.S. / IDE / etc...)
- I'm compiling on a Raspberry Pi 4 running Raspbian GNU/Linux 10 (buster)
Any clue that you can see in the CMakeCache.txt or the CMakeError.log or CMakeOutput.log files ?
- I can't see eny errors or mention of any issues.
Do you see the Pico W's LED blinking during the boot sequence (through the Green-Clock casing) ?
- When # define PICO_W is uncommented the clock show no sighs of life after being flashed. Holding the middle button when plugging in does nothing as well.
Did you update your SDK to cover the Pico W ?
- Now that's a question. I use the setup script from the document you've linked to https://github.com/raspberrypi/pico-setup/blob/master/pico_setup.sh but I'm not sure if it would miss something specific to W.
Have you tried to enter your SSID and Password on the pre-built version ? Did NTP work this way ?
- Yes and yes

**Reply by astlouys** [reply]
Hi Alex,

According to the information you provide, the problem seems related to your build environment rather than to the code itself (and the clock also works fine with the pre-built version, which excludes a problem with the product)...

I'm still wondering if your SDK covers the new Pico W. You may want to take a look at the different forums explaining how to upgrade your SDK to cover Pico W (however, I would expect some errors if your SDK is not up-to-date since some libraries would be missing in the build process).

Have you built other Pico executables with your environment ?  (without problem ?) Same question for Pico W ?

ASL

**Reply by nnmalex** [reply]
Reading through this doc https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf it seems like the default SDK does support W.

I've tried compiling with #define PICO_W commented but CMakeLists.txt for the W and it works. E.g. compiles, the display shows W as type of the controller, but NTP doesn't work off course.

**Reply by astlouys** [reply]
I would suggest that you write (or you find on the web) an "hello world" program for the Pico W. You know, this basic program that blinks the local Pico W LED.

The Pico W's LED is addressed differently than the Pico's LED and the CYW43 library (typical to Pico W) is required to blink the LED on the Pico W.

This would make sure that your build environment is OK and will allow us to put that behind and look somewhere else.

**Reply by astlouys** [reply]
Alex,

One thing that I forgot to mention in my previous mails (but mentioned in the User Guide): if / when you power-up the clock and the wi-fi credentials are incorrect (which is obviously the case the first time you run Firmware Version 9.00), it will take many long seconds (up to one minute) before you get something on the clock display. I assume you waited at least this period of time before giving up ?

**Comment by JasonWheatley** [comment]
Hi all,

Unfortunately I haven't had much time to spend on this due to other commitments. However, this is what I have tried:

My initial though was that because my main WiFi password had a '%' character in it this was causing problems with the sprintf at line 2127 in Pico-Clock-Green.c. To try and get around this I replaced sprintf with strcpy. I appended '\0' to all strings to make sure there were no buffer overflows etc. That didn't work so tried the following: 

- Erased flash completely.
- Copied in PicoW firmware that came with build checkout.
- Tried to connect to my guest WiFi which has a much simpler and shorter password than my main network.

Still it will not connect. Have tried several power cycles but still nothing.

My next step is write a small program to just connect to WiFI. Hopefully this will help me get to the bottom of the issue and then I can feedback my results.


Best regards,
Jason

**Reply by astlouys** [reply]
Hi Jason,

No problem, we all have other responsibilities to go through...

I did have another problem with "%" sign elsewhere in the code (to report % humidity on the clock display when reading from DHT22 and / or BME280). The usual way to proceed is to put 2 "%%" when you want to get one in the result string. However, if / when you make other manipulations with this "result" string (transfer or copy it), you're manipulating a string with only one "%", which is most likely to generate problems. So, it is very possible that the "%" be one of your problem. However, I don't see how it would stop the clock from starting at all. It would simply give an error while trying to "NTP sync". Do you still have "%" in your "guest WiFi network" that you tried to connect to ?

When you're able to start the Firmware, you could try to put 2 "%%" (instead of one) in your password string (assuming you don't overflow the 70 characters). Not sure if it will help. but I would give it a try.

André

**Comment by JasonWheatley** [comment]
Hi André

I have tried the %% approach but still no connect. There is no % in my guest WiFi network.

The code below based on code from [here](https://www.i-programmer.info/programming/hardware/16031-the-picow-in-c-a-better-connect.html?start=1) does connect to my WiFi (the main one with % in password). My next plan is to use some of the debug serial port facility to see what is happening and when. 

I may also get one of [these](https://thepihut.com/products/raspberry-pi-debug-probe) and do some inline debugging to try and get a handle on what is going on.

Thank you for your continued support and discussion. I am sure we will crack this!


Best regards,

Jason

```c
#include <stdio.h>
#include "pico/stdlib.h"
#include "pico/cyw43_arch.h"

int setup(
uint32_t country,
const char *ssid, 
const char *pass,
uint32_t auth)
{
  int retVal;


  if (cyw43_arch_init_with_country(country))
  {
    retVal = 1;
  }
  else
  {
    cyw43_arch_enable_sta_mode();

    if (cyw43_arch_wifi_connect_blocking(ssid, pass, auth))
    {
      retVal = 2;
    }
  }

  return retVal;
}

char ssid[] = "MySSID";
char pass[] = "MyPassword";
uint32_t country = CYW43_COUNTRY_UK;
uint32_t auth = CYW43_AUTH_WPA2_MIXED_PSK;

int main()
{
    stdio_init_all();
    setup(country, ssid, pass, auth);
    while (true)
    {
        sleep_ms(1);
    }
}
```

**Reply by astlouys** [reply]
Hi Jason,

I'm surprised that you can't connect to your guest network if there is no "%" in your SSID and / or password... In fact, I'm also surprised that the Firmware doesn't start at all !!  In the worst case, I would expect it to start, but refuse to connect to WiFi.

Harry's article is interesting. I bought his book a while ago (before he adds the Pico W part :-(  ) and he uses to go deeply enough in the explanations / examples so that you really learn something. It's definitely a good idea for you to go with a small piece of code and make it work, then return to the Green-Clock.

Good that you bought this little "Debug Probe Kit". I also bought one a few weeks ago, but didn't take the time to configure and use it so far. I would really appreciate if you can give me comments about it. My fear is... most of the "real" problem I encounter are related to timing (yes, here and there, there are a few glitches, but it's usually easy to go through them and debug them You can see many debug code in the sources, right ?!).  I doubt that the probe kit can help much with timing problems, but I may be wrong (interrupt interfering with something else... interrupts coming too fast and making an override, that kind of things). Your comments about this tool are definitely welcome.

Let me know if you find something else that could help to find the reason behind your problem.

André

**Reply by astlouys** [reply]
Jason,

I forgot to mention in my previous message (but I wrote it in the first message of this thread)... Not sure what kind of "serial port debugging" you're planning to do (apart of your plan with the Debug Probe Kit) but remember that there seems to be a random (intermittent) problem with USB CDC serial communication with the Pico W. Once the CYW43 has been initialized, sometimes, the serial output through USB becomes dead. I checked on the web, but did not find a viable solution to that problem. Keep that in mind if this is one of the steps you're planning to do...

André

**Comment by pauljcheffus** [comment]
Hi

I have exactly the same issue as the others in this discussion in that it builds ok but fails to run on the PicoW.

On power up nothing happens at all and I’ve left it for around five minutes.

if I use the precompiled Pico W version it works and if I compile the Pico version it works. It’s only when I compile the Pico W version it fails. I have uncommented the #define PICO_W line. The precompiled code is 806400 bytes but my version is 870400 bytes.

Cheers

Paul

**Reply by astlouys** [reply]
Hi Paul,

I guess the size difference of the executable may be related to the compiler / linker used and / or by some build options (release / debug / optimization settings or something else).

- Have you updated your SDK in the recent weeks to cover the Pico W (WiFi... and bluetooth support has also been released in February).
- Have you built and run some other Pico W specific executable (blinking the Pico W LED and / or using WiFi or bluetooth) ?
- Have you changed anything else in the source code beside the "#define PICO_W" ?
- What are the tools used for your development ? (IDE / compiler...)
- Absolutely no error / warning message when you build ?

I did crash the SD card of my development Raspberry Pi 4 in the last weeks and I have to rebuild my setup (fortunately, I have backups, but they are not completely up-to-date - yeah... those things happen). So, I'm working on Firmware Version 10.0 now (re-coding what I've already done some weeks ago). I'll see if I can find some clue that could explain your problem.

In the meantime, let me know if there is any other indication / clue that could help to pinpoint the problem.

André

**Reply by pauljcheffus** [reply]
Hi Andre

The environment was only built this week on a Raspberry Pi 2 model B so would expect it to be up to date and it is running Bullseye.

I wrote a program to blink the onboard LED for the Pico W and that worked fine.

I’m using whatever tools are installed by the setup script from here 
https://raw.githubusercontent.com/raspberrypi/pico-setup/master/pico_setup.sh 
and I followed the guide here to setup the environment
https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf

I also changed the DST_COUNTRY to DST_EUROPE but nothing else.

Absolutely no errors I can see and nothing happens when the compiled application is loaded to the Pico W.

Cheers

Paul

**Reply by astlouys** [reply]
Hi Paul,

The script that you gave seems to properly add / include all submodules, so I would expect that the libraries required for Pico W have been installed... And this is confirmed if you're able to blink the Pico's LED on a Pico W.

You may want to add a few "LED blinks" as the firmware starts, trying to see what goes wrong (and at which point it seems to stop / crash), or even better if you use to do it, to send some debug outputs to an external terminal emulator...

Could you confirm that the "#define RELEASE_VERSION" has not been commented-out (and that it's active) ? It is in the first 10 lines of "active code" (excluding all comment lines).

As I'm reconstructing my development environment, I'll see if I can find some clues about the problem.

André

**Reply by pauljcheffus** [reply]
Hi Andre

I will see if I can find some time to experiment over the next week. 

I can confirm I haven’t commented out the #define RELEASE_VERSION. When the build occurs it gives a warning about the build being the release version so I think that’s working correctly.

Good luck and it’s a pain when that happens.

Cheers

Paul

**Reply by pauljcheffus** [reply]
Hi Andre

This may or may not help with your investigation but I have compiled the code again (I changed the firmware version to 9.01 so I could be sure it was running the right version) and uploaded it to the Pico W. This time it works after around 5 seconds (pause while it connects to the network I assume).

The only difference between this attempt and my previous one is that the network SSID and password had been set using your precompiled version. I haven't had chance to have a proper look at the code but wondered if there was an issue on the first run and it never gets passed part of the initialisation so the network credentials set in the code don't actually get written to the flash memory. You may not see this issue as the network credentials will already be stored on your Pico W. Only issue with this theory is that your precompiled version doesn't have this issue.

Cheers

Paul

**Reply by astlouys** [reply]
Hi Paul,

Your observation is very interesting...

In theory, I made the code so that after 10 retries to connect to WiFi, the software gives up and continue its other duties (you can see the 10 tries while looking at the Pico's LED through the clock plastic case... and this is also why - the 10 retries - it takes some time to see something on the clock display) However, according to what you mention, there must be something wrong with the logic if the connection does not succeed during the power-up sequence. You're right that I may not have done "so many" tests with blank flash to trap this specific case...

What is strange is that the precompiled version should be the same as the source code I posted...

I'll check that in the next hours... In any case, I'm glad that you finally have a version that works fine ! (I do have a "Pico-Flash-Utility" to perform specific actions on the flash... like erase it and it's also available in my GitHub repositories).

André

**Reply by astlouys** [reply]
Jason, Paul, Alex,

As you know, I crashed my development system in the last days and I had to rebuild it from scratch (although I had the source code Version 9.00 for the Pico-Green-Clock).

I rebuilt the Firmware Version 9.00 and I was surprised to realize that I got the same problem that you have. After investigating, I believe I found the problem. In the file "picow_ntp_client.c", around line 472 (I may have added or removed a few lines above...) the line that begins with "ReturnCode = cyw43_arch_wifi_connect_blocking(...)"

Replace the "blocking" by "timeout_ms" and add ",1000" as the last parameter at the end of the parenthesis. Please note that I'm providing this answer soon so that you can get out of your problem, and I haven't taken the time to analyze and test this "1000 milliseconds timeout". I think it should be OK, but experience may show that it may be decreased or it should be increased.

According to what I've read on the Net, the Pico W SDK libraries have been updated in February (it seems that with the new libraries, the Pico W now supports bluetooth !) and some change in the behaviour may explain the problems that you had.

Please let me know if that solves your problem as it did for me.

André

**Reply by JasonWheatley** [reply]
Hi André,

I had a few minutes this morning and I tried your fix. Good news, it works! Thank you.

I didn't have time to debug whether the '%' character in my WiFi password would mess things up so I changed the code temporarily:

```c
    // sprintf(&FlashConfig.SSID[4],     NETWORK_NAME);
    // sprintf(&FlashConfig.Password[4], NETWORK_PASSWORD);
    strcpy(&FlashConfig.SSID[4],     NETWORK_NAME);
    strcpy(&FlashConfig.Password[4], NETWORK_PASSWORD);
```

I will re-visit this later and let you know the results.

Thank you once again for the fix and have a happy Easter.


Best regards,

Jason

**Reply by pauljcheffus** [reply]
Hi Andre

I cleared the flash and tried it again. It does now indeed work as expected, however it seems to ignore the SSID and password I put in the code before compiling. I had to enter them manually via the procedure in the manual, not a real issue for me as my credentials aren’t that long but could be for others.

I may of course misunderstood what I needed to do but it did seem straightforward.

Cheers

Paul

**Reply by astlouys** [reply]
Hi Paul,

Glad to know that it also solves your problem !

In the main source file (Pico-Green-Clock.c), near line 287 (since I do work in the code, the line number may be different for you), you should see:
#define NETWORK_NAME   "MyNetworkName"
#define NETWORK_PASSWORD  "MyPassword"

You should replace both item between between "double-quotes" with your own SSID and password (while keeping the surrounding "double-quotes") and it should work. As Jason mentioned, however, if you have strange characters in either your SSID and / or password, you may encounter some problem (characters that have a special signification in C language like "slash", "back-slash", "percent-sign", etc...  and this, even if your SSID or password is short).

Whatever the case, if it works fine now, the best is to leave it this way, except if you're really curious (smile).

André

**Reply by astlouys** [reply]
I forgot to mention: If you already ran Version 9.00 before (I think you did), you should instead look around line 2272...
Copy-paste the 2 lines:
sprintf(&FlashConfig.SSID[4], NETWORK_NAME);
sprintf(&FlashConfig.NETWORK_PASSWORD);

First, replace NETWORK_NAME and NETWORK_PASSWORD with your own (between "double-quotes") if you haven't already done it around line 287 as indicated in my previous message...

Paste the 2 lines just under the commented line:
/*** One-time FlashConfig writes may be inserted below... ***/

(this is also around line 2282).

**Reply by pauljcheffus** [reply]
Hi Andre

I wiped the flash so it should have worked with changing the code around line 287 which is what I did, however I’ve realised what I did wrong. I didn’t wait for the first scroll message which would have written it to the flash.

Thanks for all your help in getting this working for me. I can now experiment with the clock.

Cheers

Paul

**Reply by astlouys** [reply]
Eh, eh... I realize that you begin to understand / know the product quite well now... ;-)
